addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,586 @@
+		return nil, err
+	}
+
+	tanzuClusterBootstrap := &runtanzuv1alpha3.TanzuClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), tanzuClusterBootstrap)
+	// if found and resolved tkr is the same, return found object as the TKR is supposed to be immutable
+	// also preserves any user changes
+	if err == nil && tkrName == tanzuClusterBootstrap.Status.ResolvedTKR {
+		return tanzuClusterBootstrap, nil
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if tanzuClusterBootstrap.UID == "" {
+		log.Info("tanzuClusterBootstrap for cluster does not exist, cloning from template")
+
+		tanzuClusterBootstrap.Name = cluster.Name
+		tanzuClusterBootstrap.Namespace = cluster.Namespace
+		tanzuClusterBootstrap.Spec = tanzuClusterBootstrapTemplate.Spec.DeepCopy()
+		secrets, providers, err := r.cloneSecretsAndProviders(cluster, tanzuClusterBootstrap, tanzuClusterBootstrapTemplate.Namespace, log)
+		if err != nil {
+			r.Log.Error(err, "unable to clone secrets, providers")
+			return nil, err
+		}
+
+		tanzuClusterBootstrap.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+				Kind:               cluster.Kind,
+				Name:               cluster.Name,
+				UID:                cluster.UID,
+				Controller:         pointer.BoolPtr(true),
+				BlockOwnerDeletion: pointer.BoolPtr(true),
+			},
+		}
+
+		if err := r.Client.Create(r.context, tanzuClusterBootstrap); err != nil {
+			return nil, err
+		}
+		// ensure ownerRef of tanzuClusterBootstrap on created secrets and providers, this can only be done after
+		// tanzuClusterBootstrap is created
+		ownerRef := metav1.OwnerReference{
+			APIVersion:         runtanzuv1alpha3.GroupVersion.String(),
+			Kind:               "TanzuClusterBootstrap", // kind is empty after create
+			Name:               tanzuClusterBootstrap.Name,
+			UID:                tanzuClusterBootstrap.UID,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}
+		if err := r.ensureOwnerRef(&ownerRef, secrets, providers); err != nil {
+			r.Log.Error(err, "unable to ensure ownerref on created secrets and providers", "tanzuclusterbootstrap", tanzuClusterBootstrap)
+			return nil, err
+		}
+
+		tanzuClusterBootstrap.Status.ResolvedTKR = tkrName
+		if err := r.Status().Update(r.context, tanzuClusterBootstrap); err != nil {
+			return nil, err
+		}
+		r.Log.Info("cloned tanzuclusterbootstrap", "tanzuclusterbootstrap", tanzuClusterBootstrap)
+		return tanzuClusterBootstrap, nil
+	}
+
+	// TODO upgrade needs patch (update versions of all packages, merge configs, add additional packages and remove packages that don't exist anymore)
+	// https://github.com/vmware-tanzu/tanzu-framework/issues/1584
+	if tkrName != tanzuClusterBootstrap.Status.ResolvedTKR {
+		log.Info("TODO handle upgrade")
+		return nil, nil
+	}
+	return nil, errors.New("should not happen")
```

Comment:
nit: In theory that our code should not reach this line, but just in case if it does, do we want to provide some meaningful error message ?